File: server/lib/rate-limit.ts
import rateLimit from 'express-rate-limit';
import slowDown from 'express-slow-down';
import { Request, Response, NextFunction } from 'express';

// In-memory store for tracking abuse (in production, use Redis)
interface AbuseTracker {
  ip: Map<string, { count: number; firstSeen: number; blocked: boolean }>;
  user: Map<string, { count: number; firstSeen: number; blocked: boolean }>;
}

const abuseTracker: AbuseTracker = {
  ip: new Map(),
  user: new Map(),
};

// Clean up old entries every 5 minutes
setInterval(() => {
  const now = Date.now();
  const oneHour = 60 * 60 * 1000;
  
  for (const [key, value] of abuseTracker.ip.entries()) {
    if (now - value.firstSeen > oneHour) {
      abuseTracker.ip.delete(key);
    }
  }
  
  for (const [key, value] of abuseTracker.user.entries()) {
    if (now - value.firstSeen > oneHour) {
      abuseTracker.user.delete(key);
    }
  }
}, 5 * 60 * 1000);

/**
 * Get client IP address from request
 */
function getClientIp(req: Request): string {
  const forwarded = req.headers['x-forwarded-for'];
  if (typeof forwarded === 'string') {
    return forwarded.split(',')[0].trim();
  }
  return req.ip || req.socket.remoteAddress || 'unknown';
}

/**
 * Get user identifier from request
 */
function getUserId(req: Request): string | null {
  return (req.headers['x-privy-user-id'] as string) || null;
}

// Abuse thresholds
const ABUSE_THRESHOLD_COUNT = 50; // Block after 50 failed requests
const ABUSE_WINDOW_MS = 15 * 60 * 1000; // 15 minutes

/**
 * Track failed requests for abuse detection
 */
export function trackFailedRequest(req: Request, reason: string) {
  const ip = getClientIp(req);
  const userId = getUserId(req);
  const now = Date.now();
  
  // Track by IP
  const ipEntry = abuseTracker.ip.get(ip) || { count: 0, firstSeen: now, blocked: false };
  
  // Reset count if window has passed
  if (now - ipEntry.firstSeen > ABUSE_WINDOW_MS) {
    ipEntry.count = 0;
    ipEntry.firstSeen = now;
  }
  
  ipEntry.count += 1;
  abuseTracker.ip.set(ip, ipEntry);
  
  // Auto-block if threshold exceeded
  if (ipEntry.count >= ABUSE_THRESHOLD_COUNT && !ipEntry.blocked) {
    ipEntry.blocked = true;
    console.warn(`🚫 Auto-blocked IP ${ip} after ${ipEntry.count} failed requests`);
  }
  
  // Track by user if available
  if (userId) {
    const userEntry = abuseTracker.user.get(userId) || { count: 0, firstSeen: now, blocked: false };
    
    // Reset count if window has passed
    if (now - userEntry.firstSeen > ABUSE_WINDOW_MS) {
      userEntry.count = 0;
      userEntry.firstSeen = now;
    }
    
    userEntry.count += 1;
    abuseTracker.user.set(userId, userEntry);
    
    // Auto-block if threshold exceeded
    if (userEntry.count >= ABUSE_THRESHOLD_COUNT && !userEntry.blocked) {
      userEntry.blocked = true;
      console.warn(`🚫 Auto-blocked user ${userId} after ${userEntry.count} failed requests`);
    }
  }
}

/**
 * Check if IP or user is blocked
 */
export function isBlocked(req: Request): { blocked: boolean; reason?: string } {
  const ip = getClientIp(req);
  const userId = getUserId(req);
  
  // Check IP block
  const ipEntry = abuseTracker.ip.get(ip);
  if (ipEntry?.blocked) {
    return { blocked: true, reason: 'IP address is blocked due to abuse' };
  }
  
  // Check user block
  if (userId) {
    const userEntry = abuseTracker.user.get(userId);
    if (userEntry?.blocked) {
      return { blocked: true, reason: 'User account is blocked due to abuse' };
    }
  }
  
  return { blocked: false };
}

/**
 * Block an IP or user
 */
export function blockClient(req: Request, durationMs: number = 60 * 60 * 1000) {
  const ip = getClientIp(req);
  const userId = getUserId(req);
  
  const ipEntry = abuseTracker.ip.get(ip) || { count: 0, firstSeen: Date.now(), blocked: false };
  ipEntry.blocked = true;
  abuseTracker.ip.set(ip, ipEntry);
  
  if (userId) {
    const userEntry = abuseTracker.user.get(userId) || { count: 0, firstSeen: Date.now(), blocked: false };
    userEntry.blocked = true;
    abuseTracker.user.set(userId, userEntry);
  }
  
  // Auto-unblock after duration
  setTimeout(() => {
    const ipEntry = abuseTracker.ip.get(ip);
    if (ipEntry) {
      ipEntry.blocked = false;
    }
    if (userId) {
      const userEntry = abuseTracker.user.get(userId);
      if (userEntry) {
        userEntry.blocked = false;
      }
    }
  }, durationMs);
}

/**
 * Abuse detection middleware - blocks clients with too many failed requests
 */
export function abuseDetectionMiddleware(req: Request, res: Response, next: NextFunction) {
  const blocked = isBlocked(req);
  if (blocked.blocked) {
    return res.status(403).json({
      error: 'Access denied',
      details: blocked.reason,
    });
  }
  
  next();
}

/**
 * General API rate limiter - IP-based
 */
export const generalLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 100, // Limit each IP to 100 requests per windowMs
  message: {
    error: 'Too many requests',
    details: 'Please try again later',
  },
  standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
  legacyHeaders: false, // Disable the `X-RateLimit-*` headers
  skip: (req) => {
    // Skip rate limiting for health checks
    return req.path === '/health';
  },
  handler: (req, res) => {
    trackFailedRequest(req, 'rate_limit_exceeded');
    res.status(429).json({
      error: 'Too many requests',
      details: 'Rate limit exceeded. Please try again later.',
    });
  },
});

/**
 * Strict rate limiter for sensitive endpoints (payments, etc.)
 */
export const strictLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 20, // Limit each IP to 20 requests per windowMs
  message: {
    error: 'Too many requests',
    details: 'Rate limit exceeded for this endpoint',
  },
  standardHeaders: true,
  legacyHeaders: false,
  handler: (req, res) => {
    trackFailedRequest(req, 'strict_rate_limit_exceeded');
    res.status(429).json({
      error: 'Too many requests',
      details: 'Rate limit exceeded for this sensitive endpoint. Please try again later.',
    });
  },
});

/**
 * User-based rate limiter (if user is authenticated)
 */
export const userLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 200, // Limit each user to 200 requests per windowMs
  keyGenerator: (req) => {
    const userId = getUserId(req);
    return userId || getClientIp(req); // Fallback to IP if no user
  },
  message: {
    error: 'Too many requests',
    details: 'User rate limit exceeded',
  },
  standardHeaders: true,
  legacyHeaders: false,
  handler: (req, res) => {
    trackFailedRequest(req, 'user_rate_limit_exceeded');
    res.status(429).json({
      error: 'Too many requests',
      details: 'User rate limit exceeded. Please try again later.',
    });
  },
});

/**
 * Slow down middleware - gradually increases response time for repeated requests
 */
export const speedLimiter = slowDown({
  windowMs: 15 * 60 * 1000, // 15 minutes
  delayAfter: 50, // Allow 50 requests per windowMs without delay
  delayMs: 100, // Add 100ms delay per request after delayAfter
  maxDelayMs: 2000, // Maximum delay of 2 seconds
  skip: (req) => {
    return req.path === '/health';
  },
});

/**
 * Request size limiter middleware
 */
export function requestSizeLimiter(maxSizeBytes: number = 1024 * 1024) { // Default 1MB
  return (req: Request, res: Response, next: NextFunction) => {
    const contentLength = parseInt(req.headers['content-length'] || '0', 10);
    
    if (contentLength > maxSizeBytes) {
      trackFailedRequest(req, 'request_too_large');
      return res.status(413).json({
        error: 'Request too large',
        details: `Request body exceeds maximum size of ${maxSizeBytes / 1024}KB`,
      });
    }
    
    next();
  };
}

File: mcp-server/app/core/rate_limit.py
"""
Rate limiting and abuse protection for FastAPI MCP server.
"""

from typing import Optional
from fastapi import Request, HTTPException, status
from slowapi import Limiter
from slowapi.util import get_remote_address
from slowapi.errors import RateLimitExceeded
from slowapi.middleware import SlowAPIMiddleware
import structlog

logger = structlog.get_logger(__name__)

# Abuse thresholds
ABUSE_THRESHOLD_COUNT = 50  # Block after 50 failed requests
ABUSE_WINDOW_SECONDS = 15 * 60  # 15 minutes

# In-memory store for abuse tracking (in production, use Redis)
_abuse_tracker: dict[str, dict] = {
    "ip": {},
    "user": {},
}

# Initialize rate limiter
limiter = Limiter(key_func=get_remote_address)


def get_user_id(request: Request) -> Optional[str]:
    """Extract user ID from request headers."""
    return request.headers.get("X-Privy-User-Id") or request.headers.get("X-User-Id")


def get_client_ip(request: Request) -> str:
    """Get client IP address from request."""
    forwarded = request.headers.get("X-Forwarded-For")
    if forwarded:
        return forwarded.split(",")[0].strip()
    return get_remote_address(request)


def track_failed_request(request: Request, reason: str) -> None:
    """Track failed requests for abuse detection."""
    import time
    
    ip = get_client_ip(request)
    user_id = get_user_id(request)
    now = time.time()
    
    # Track by IP
    if ip not in _abuse_tracker["ip"]:
        _abuse_tracker["ip"][ip] = {"count": 0, "first_seen": now, "blocked": False}
    
    ip_entry = _abuse_tracker["ip"][ip]
    
    # Reset count if window has passed
    if now - ip_entry["first_seen"] > ABUSE_WINDOW_SECONDS:
        ip_entry["count"] = 0
        ip_entry["first_seen"] = now
    
    ip_entry["count"] += 1
    
    # Auto-block if threshold exceeded
    if ip_entry["count"] >= ABUSE_THRESHOLD_COUNT and not ip_entry["blocked"]:
        ip_entry["blocked"] = True
        logger.warn("auto_blocked_ip", ip=ip, count=ip_entry["count"])
    
    # Track by user if available
    if user_id:
        if user_id not in _abuse_tracker["user"]:
            _abuse_tracker["user"][user_id] = {"count": 0, "first_seen": now, "blocked": False}
        
        user_entry = _abuse_tracker["user"][user_id]
        
        # Reset count if window has passed
        if now - user_entry["first_seen"] > ABUSE_WINDOW_SECONDS:
            user_entry["count"] = 0
            user_entry["first_seen"] = now
        
        user_entry["count"] += 1
        
        # Auto-block if threshold exceeded
        if user_entry["count"] >= ABUSE_THRESHOLD_COUNT and not user_entry["blocked"]:
            user_entry["blocked"] = True
            logger.warn("auto_blocked_user", user_id=user_id, count=user_entry["count"])
    
    logger.warn(
        "abuse_tracked",
        ip=ip,
        user_id=user_id,
        reason=reason,
        ip_count=ip_entry["count"],
        user_count=_abuse_tracker["user"].get(user_id, {}).get("count", 0) if user_id else 0,
    )


def is_blocked(request: Request) -> tuple[bool, Optional[str]]:
    """Check if IP or user is blocked."""
    ip = get_client_ip(request)
    user_id = get_user_id(request)
    
    # Check IP block
    ip_entry = _abuse_tracker["ip"].get(ip)
    if ip_entry and ip_entry.get("blocked"):
        return True, "IP address is blocked due to abuse"
    
    # Check user block
    if user_id:
        user_entry = _abuse_tracker["user"].get(user_id)
        if user_entry and user_entry.get("blocked"):
            return True, "User account is blocked due to abuse"
    
    return False, None


def block_client(request: Request, duration_seconds: int = 3600) -> None:
    """Block an IP or user."""
    ip = get_client_ip(request)
    user_id = get_user_id(request)
    
    if ip not in _abuse_tracker["ip"]:
        _abuse_tracker["ip"][ip] = {"count": 0, "blocked": False}
    _abuse_tracker["ip"][ip]["blocked"] = True
    
    if user_id:
        if user_id not in _abuse_tracker["user"]:
            _abuse_tracker["user"][user_id] = {"count": 0, "blocked": False}
        _abuse_tracker["user"][user_id]["blocked"] = True
    
    logger.warn("client_blocked", ip=ip, user_id=user_id, duration=duration_seconds)
    
    # Auto-unblock after duration (simplified - in production use Redis with TTL)
    import asyncio
    async def unblock_after_delay():
        await asyncio.sleep(duration_seconds)
        if ip in _abuse_tracker["ip"]:
            _abuse_tracker["ip"][ip]["blocked"] = False
        if user_id and user_id in _abuse_tracker["user"]:
            _abuse_tracker["user"][user_id]["blocked"] = False
    
    asyncio.create_task(unblock_after_delay())


async def abuse_detection_middleware(request: Request, call_next):
    """Middleware to check if client is blocked."""
    blocked, reason = is_blocked(request)
    if blocked:
        logger.warn("request_blocked", ip=get_client_ip(request), user_id=get_user_id(request), reason=reason)
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"error": "Access denied", "details": reason},
        )
    
    response = await call_next(request)
    
    # Track failed requests (4xx and 5xx status codes)
    if response.status_code >= 400:
        track_failed_request(request, f"http_{response.status_code}")
    
    return response


# Rate limit decorators
def general_limiter():
    """General API rate limiter - 100 requests per 15 minutes."""
    return limiter.limit("100/15minutes")


def strict_limiter():
    """Strict rate limiter for sensitive endpoints - 20 requests per 15 minutes."""
    return limiter.limit("20/15minutes")


def user_limiter():
    """User-based rate limiter - 200 requests per 15 minutes."""
    def key_func(request: Request) -> str:
        user_id = get_user_id(request)
        return user_id or get_client_ip(request)
    
    return limiter.limit("200/15minutes", key_func=key_func)


# Custom exception handler for rate limit exceeded
def rate_limit_handler(request: Request, exc: RateLimitExceeded):
    """Custom handler for rate limit exceeded."""
    track_failed_request(request, "rate_limit_exceeded")
    logger.warn(
        "rate_limit_exceeded",
        ip=get_client_ip(request),
        user_id=get_user_id(request),
        limit=exc.detail,
    )
    raise HTTPException(
        status_code=status.HTTP_429_TOO_MANY_REQUESTS,
        detail={
            "error": "Too many requests",
            "details": "Rate limit exceeded. Please try again later.",
            "retry_after": exc.retry_after,
        },
    )
